### PR TITLE
The managed array style for RingBuffer performance.

### PR DESCRIPTION
**Pull request by michaelbpaulson**
Instead of using `push` and `shift` a simple modulo is used to determine the position to insert.  This does cause an issue as `this.records[0]` is not the oldest record.

Instead a `get` method has been added to help get records by index.

### CODE REVIEW

**Comment by michaelbpaulson** [comment]
#279

**Comment by michaelbpaulson** [comment]
If `size()` is not wanted a simple getter for length could do as well.  
Second if limit is expected to be updatable then a getter/setter will be required.

**Comment by yunong** [comment]
@michaelbpaulson this looks great. I would add a  ```getLatest()``` API, to make it easy to query the latest record.

File: lib/RingBuffer.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    this.limit = options && options.limit ? options.limit : 100;
+    this.writable = true;
+    this.records = new Array(this.limit);
+    this._idx = -1;
+    EventEmitter.call(this);
+};
+
+util.inherits(RingBuffer, EventEmitter);
+
+RingBuffer.prototype.write = function (record) {
```
Comment (yunong) [inline lib/RingBuffer.js]: You'll definitely want to name your functions here, to make it easier to find them in core dumps.

e.g.

```js
RingBuffer.prototype.write = function write (record) {
...
```

**Comment by michaelbpaulson** [comment]
@yunong I added the `getLatest` function.  The only problem i see is 

```javascript
// inverse is true.
get(-1) === getLatest(0) 
```

Should i disallow -1 keys or keep the api the same?

**Comment by michaelbpaulson** [comment]
@yunong anything I need to do for this?

**Comment by yunong** [comment]
@trentm Can you take a look? If it's okay would you mind merging?

**Comment by michaelbpaulson** [comment]
It does break contract so it will require a minor version bump when next tag is emitted :)

**Comment by trentm** [comment]
Some comments. Generally looks good, though a few things because this is a compatibility change:


- What is the speed diff? Is there any data to quote here? If only a modest
  gain then I'm more skeptical :)

- Looking at restify's RequestCaptureStream updates for this:
    https://github.com/restify/node-restify/blob/master/lib/bunyan_helper.js#L114-L179
  Is there a coming change to update that usage? That's the major user that
  I know of.
  IIUC, one of the common use cases for the RingBuffer is to iterate through
  all the items. Is there a better way to iterate the items in the ring than
  calling `.get()` N times?

- doc/comment on the _idx field would be nice

- Please put a try/catch around the RingBuffer `require` so that it is
  optional. IOW, one can use `bunyan.js` standalone

- This will be a major ver bump for possible backward incompat change. I don't
  know if that is not the norm, but I use a major ver bump whenever I want
  to signal to users that they might *have* to change their code to update.

- Please change `this.records` to `this._records` to be explicit about that
  being internal field now.

**Comment by trentm** [comment]
Now I'm thinking of having:

- lib/ringbuffer.js  which is the current slow RingBuffer and importable via a getter from `bunyan.RingBuffer`. That getter would throw if "lib/ringbuffer.js" isn't available
- this new guy becomes lib/fastringbuffer.js or something. Then we maintain the old functionality and don't have to worry about all this compat dancing.
- worth thinking about putting these in `lib/streams/*.js` subdir.

**Comment by michaelbpaulson** [comment]
That sounds good @trentm .  I can take care of that here shortly.

**Comment by yunong** [comment]
What are the thoughts around making ringbuffer a separate module?

On Monday, August 17, 2015, Trent Mick <notifications@github.com> wrote:

> Now I'm thinking of having:
>
>    - lib/ringbuffer.js which is the current slow RingBuffer and
>    importable via a getter from bunyan.RingBuffer. That getter would
>    throw if "lib/ringbuffer.js" isn't available
>    - this new guy becomes lib/fastringbuffer.js or something. Then we
>    maintain the old functionality and don't have to worry about all this
>    compat dancing.
>    - worth thinking about putting these in lib/streams/*.js subdir.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/trentm/node-bunyan/pull/280#issuecomment-132038850>.
>


-- 
Via mobile

**Comment by trentm** [comment]
@yunong having a separate module is totally cool. There is no reason it needs to be part of bunyan core. It might still be nice to use a differentiating name to not have two things called exactly "RingBuffer".

**Comment by michaelbpaulson** [comment]
@yunong @trentm Sorry for the dangling PR.  that always sucks.  I am going to move the RingBuffer to an NPM module.  If bunyan wants it, they can include it.

**Comment by michaelbpaulson** [comment]
@trentm I currently have it here https://github.com/michaelbpaulson/BunyanRingBuffer.  If you would like it moved somewhere, please let me know and I can give you the repo / npm.

**Pull request closed without merging**